### DISCUSSION TITLE
Getting secondary guesses

### DISCUSSION THREAD

**Discussion by fluffytid** [ultralytics/ultralytics#10180] [Q&A]
I am working with playing card recognition.  I am using a YOLO model (carddetector.pt)

The YOLO I have trained is quite good detecting where there is a playing card, but it can still improve when detecting which card it is.

Sometimes when it doubts between 2 similar classes/cards (such as a 7 where fingers of players hide the middle icon making it look like a 6), some frames will give me one class, while the others will give me the other. On rare cases it will give me both. Those are fine for me.

The problem comes when it consistently guesses the wrong card over a long period. I would rather know that that 0.93 confidence class had actually a strong competitior on the 0.80s than not knowing it. But I have found no way to gather alternative classes. Is there any?

**Comment by pderrenger** [comment]
Hello! Your project on playing card recognition using YOLO sounds intriguing! 🎴

To tackle your issue of wanting to know secondary guesses for classes where the model is confident but possibly wrong, you could consider lowering the confidence threshold (`conf`) and then manually handling the multiple detections. This approach enables you to see alternative classes that are also considered likely by the model but with lower confidence scores.

For instance, when running predictions, you could set a lower threshold:
```python
model.predict(source='path/to/image.jpg', conf=0.5)
```

This code would give you detections with a confidence score above 0.5, including those that weren't the top pick but are still significantly confident. You'd then need to filter these detections based on your requirements, possibly considering the top 2 or 3 for your analysis.

Just keep in mind that lowering the confidence threshold might result in more detections per image, including potentially irrelevant ones, so you'll need to balance this based on your application's needs.

Hope this helps, and best of luck with your card detection efforts! 🍀

**Comment by fluffytid** [comment]
I am not sure I am using predictions or a different method nor the
difference between them. My current line is this:

results = model(img, stream=True, save_conf=True)

**Reply by glenn-jocher** [reply]
Hey! It looks like you're on the right track with your YOLO implementation for playing card recognition. Lowering the confidence threshold, as suggested, is indeed a practical way to explore secondary class guesses. The line you mentioned,

```python
results = model(img, stream=True, save_conf=True)
```
seems correctly set for processing sequences or streams of images. To integrate the approach of capturing lower-confidence detections (secondary guesses), you just need to adjust the confidence threshold when you run predictions:

```python
results = model.predict(source='path/to/image.jpg', conf=0.5, save_conf=True)
```
This would indeed give you detections over the specified confidence level, allowing you to examine alternative guesses more closely. Adjust the `conf` value according to the sensitivity you need. 

Just ensure your post-processing handles the extra detections efficiently, focusing on balancing between capturing potentially missed correct detections and filtering out irrelevant ones. Happy coding and best wishes on improving your card detection model! 🌟
